freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtRoleComponent.java

```diff
@@ -53,7 +53,18 @@ public void addRoleAndPrivileges(Optional<Service> service, Optional<Host> host,
         if (roleRequest != null && StringUtils.isNotBlank(roleRequest.getRoleName())) {
             Set<Role> allRole = ipaClient.findAllRole();
             Optional<Role> optionalRole = allRole.stream().filter(role -> role.getCn().equals(roleRequest.getRoleName())).findFirst();
-            Role role = optionalRole.isPresent() ? optionalRole.get() : ipaClient.addRole(roleRequest.getRoleName());
+            Role role;
```

Comment:
logic is getting more and more complex here, we shall add a bit more test coverage, focusing also on the edge cases

Reply:
I added some tests for the race condition and also for some of the prior corner cases that were previously insufficiently unit tested.